### PR TITLE
Enforces running SPK Project commands in order and Lifecycle pipeline refactoring to use variable groups from Bedrock.yaml

### LINKED ISSUE #886: SPK Project commands must run in order
**As a:** SPK user and developer

**I want:** the `spk project ...` commands to have a designated order in which
they are called to be run.

**So that:** their is no ambiguity in ways the commands can be run and the order
in which they must be called.

**Describe the solution you'd like:**

- [x] `spk project ...` commands needs to run in order -- Each command should
      make sure it’s run in order. throw errors/info messages to user if not.
  - [x] `init` - can be run anytime -- existing functionality should be fine as
        it currently logs a warning that an existing `bedrock.yaml` file exists
        if the project has already been initialized and performs a noop
  - [x] `create-variable-group` - checks to see an existing `bedrock.yaml` file
        exists and adds the variable group to it; throws if it does not.
  - [x] `install-lifecycle-pipeline` - checks to see that a `bedrock.yaml` file
        exists and that it has at least one variable group; throws if it does
        not
- [x] Refactor `install-lifecycle-pipeline` to not use baked in variables in the
      script any longer but instead only consume the variable groups created via
      `create-variable-group`

**Acceptance Criteria:**

Refer to above task list

**Describe alternatives you've considered:**

**Additional context:**

Currently `create-variable-group` and `install-lifecycle-pipeline` are disjoint
commands. The pipeline created via `install-lifecycle-pipeline` expects to
receive all the parameters contained in the variable group via command flags and
does not consume the variable group create via `cvg`. The only way to consume
the variable group is to add it in manually.

**Does this require updates to documentation?:**

Yes

### PR DESCRIPTION
**Pull request by sarath-p**
Closes https://github.com/microsoft/bedrock/issues/886

### CODE REVIEW

File: src/types.d.ts [hunk 1]
```diff
@@ -143,7 +143,9 @@ export interface IVariableGroupData {
   name: string;
   description: string;
   type: string;
-  variables: IVariableGroupDataVariable[];
+  variables: {
```
Comment (dennisseah) [inline src/types.d.ts]: can we have it as 
```
   variables: IVariableGroupDataVariable
```

File: src/lib/pipelines/variableGroup.ts [hunk 2]
```diff
@@ -267,9 +267,9 @@ export interface IVariablesMap {
  * @returns `IVariablesMap[]` with Varibale Group variables
  */
 export const buildVariablesMap = async (
-  variables: IVariablesMap[]
-): Promise<IVariablesMap> => {
-  const variablesMap: IVariablesMap = {};
+  variables: IVariableGroupDataVariable
```
Comment (dennisseah) [inline src/lib/pipelines/variableGroup.ts]: can you remove `IVariablesMap` interface declaration in this file?

File: src/lib/pipelines/variableGroup.test.ts [hunk 3]
```diff
@@ -406,96 +405,85 @@ describe("authorizeAccessToAllPipelines", () => {
...
   });
 
   test("should create empty variable map with no secrets", async () => {
-    const variables: IVariablesMap[] = [];
-    let secretsMap: IVariablesMap | undefined;
-    const expectedMap: any = `{}`;
+    const variables: IVariableGroupDataVariable = {};
+    let secretsMap: IVariableGroupDataVariable | undefined;
+    const expectedMap = `{}`;
     secretsMap = await buildVariablesMap(variables);
```
Comment (dennisseah) [inline src/lib/pipelines/variableGroup.test.ts]: prefer to be like this
```
const variables: IVariableGroupDataVariable = {};
const secretsMap = await buildVariablesMap(variables);
```
typescript is able to infer the type for `secretsMap`

File: src/lib/pipelines/variableGroup.test.ts [hunk 4]
```diff
@@ -406,96 +405,85 @@ describe("authorizeAccessToAllPipelines", () => {
...
 
-    const variables: IVariablesMap[] = [];
-    variables.push(var1);
-    let secretsMap: IVariablesMap | undefined;
-    const expectedMap: any = `{"0":{"secret1":{"enabled":true}}}`;
+    let secretsMap: IVariableGroupDataVariable | undefined;
+    // const expectedMap: any = `{{"secret1":{"enabled":true}}}`;
+    const expectedMap = '{"secret1":{"enabled":true}}';
     secretsMap = await buildVariablesMap(variables);
     expect(JSON.stringify(secretsMap)).toEqual(expectedMap);
```
Comment (dennisseah) [inline src/lib/pipelines/variableGroup.test.ts]: can remove
`let secretsMap: IVariableGroupDataVariable | undefined;` 
and do
`const secretsMap = await buildVariablesMap(variables);`
typescript is able to infer the type for secretsMap

File: src/lib/pipelines/variableGroup.test.ts [hunk 5]
```diff
@@ -406,96 +405,85 @@ describe("authorizeAccessToAllPipelines", () => {
...
-    variables.push(var1);
-    variables.push(var2);
-
-    let secretsMap: IVariablesMap | undefined;
-    const expectedMap: any = `{"0":{"secret1":{"enabled":false}},"1":{"secret2":{"enabled":true}}}`;
+    let secretsMap: IVariableGroupDataVariable | undefined;
+    // const expectedMap = `{{"secret1":{"enabled":false}},"1":{"secret2":{"enabled":true}}}`;
+    const expectedMap =
+      '{"secret1":{"enabled":false},"secret2":{"enabled":true}}';
     secretsMap = await buildVariablesMap(variables);
```
Comment (dennisseah) [inline src/lib/pipelines/variableGroup.test.ts]: remove `let secretsMap: IVariableGroupDataVariable | undefined;`
and have `const secretsMap = await buildVariablesMap(variables);`
typescript is able to infer the type for secretsMap

File: src/lib/pipelines/variableGroup.ts (hunk 2 above)
Comment (sarath-p) [reply src/lib/pipelines/variableGroup.ts]: Good catch.

File: src/types.d.ts (hunk 1 above)
Comment (sarath-p) [reply src/types.d.ts]: Yes!

**Comment by andrebriggs** [comment]
@sarath-p any update on the feedback here?

File: src/lib/pipelines/variableGroup.test.ts (hunk 3 above)
Comment (sarath-p) [reply src/lib/pipelines/variableGroup.test.ts]: Agreed! I think it was copy/paste previously.

File: src/types.d.ts (hunk 1 above)
Comment (sarath-p) [reply src/types.d.ts]: Changed it.

**Review by dennisseah — APPROVED** [review]
LGTM

**Review by mtarng — APPROVED** [review]

**Review by dennisseah — APPROVED** [review]

File: src/commands/project/create-variable-group.test.ts [hunk 6]
```diff
@@ -375,6 +378,76 @@ describe("updateLifeCyclePipeline", () => {
     }
     expect(noFileError).toBeDefined();
   });
+
+  test("Should pass adding variable groups when bedrock file exists with empty variableGroups", async () => {
```
Comment (mtarng) [inline src/commands/project/create-variable-group.test.ts:382]: @sarath-p do these two tests means "we should NOT add variable groups when a bedrock file already exists with an empty variable group?" and "we should NOT add variable groups when a bedrock file already exists with one variable group?" as far as i know, we should support adding multiple variable groups and add the list of them to each pipeline.

@bnookala @evanlouie

**Review by mtarng — CHANGES_REQUESTED** [review]

File: src/commands/project/create-variable-group.test.ts (hunk 6 above)
Comment (sarath-p) [reply src/commands/project/create-variable-group.test.ts:382]: No, it always replaces variable groups in the lifecycle pipeline from the list of variable groups in bedrock.yaml. The tests are simply checking whether code works when empty array and array with one variable group present in bedrock.yaml.

**Review by mtarng — APPROVED** [review]

**Review by evanlouie — APPROVED** [review]
lgtm 👍

**Pull request merged**